syncs/syncs.go

```diff
@@ -192,6 +192,22 @@ func (m *Map[K, V]) LoadOrStore(key K, value V) (actual V, loaded bool) {
 	return actual, loaded
 }
 
+// LoadOrInit returns the value for the given key if it exists, otherwise f is
+// called to construct the value to be set.
+func (m *Map[K, V]) LoadOrInit(key K, f func() V) (actual V, initialized bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if actual, ok := m.m[key]; ok {
+		return actual, ok
+	}
+
+	initialized = true
+	actual = f()
```

Comment:
Should we hold `m.mu.Lock` while `f` is running?

Reply:
Benefit of holding it:
* We guarantee that `f` is only ever called once if the key is not present.

Detriment of holding it:
* We could hold up the map for longer than necessary, especially of `f` is slow.

Reply:
The answer is in fact fairly simple: we should hold it because that is the only version that an external implementor can not implement.

Reply:
I'm sold.

Reply:
Document this guarantee, though.